File: server/audit/signals.py
"""Gather SpecBox-specific signals for the Maintainability 60/40 mix and
related characteristics.

Reads directly from the project's `.quality/` tree and from the engine state
registry when available. Never raises — returns empty dicts on failure.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def fetch_specbox_signals(project_path: Path, project_name: str, state_path: Path | None = None) -> dict[str, Any]:
    signals: dict[str, Any] = {
        "ac_status": {"total": 0, "completed": 0},
        "evidence": {"uc_total": 0, "uc_with_evidence": 0},
        "healing": {"total_events": 0, "resolved": 0, "failed": 0},
        "board": {"us_total": 0, "us_blocked": 0},
        "acceptance": {"accepted": 0, "rejected": 0},
        "tests": {"total": 0, "passed": 0},
        "prd_divergence_ratio": 0.0,
    }

    quality_dir = project_path / ".quality"
    if not quality_dir.exists():
        return signals

    # Evidence + healing from .quality/evidence/*/
    evidence_dir = quality_dir / "evidence"
    if evidence_dir.is_dir():
        ucs = [d for d in evidence_dir.iterdir() if d.is_dir()]
        signals["evidence"]["uc_total"] = len(ucs)
        with_evidence = 0
        total_healing = 0
        resolved = 0
        failed = 0
        for uc in ucs:
            has_any = any(f.is_file() for f in uc.iterdir())
            if has_any:
                with_evidence += 1
            healing_file = uc / "healing.jsonl"
            if healing_file.exists():
                for line in healing_file.read_text(encoding="utf-8", errors="ignore").splitlines():
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        event = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    total_healing += 1
                    if event.get("result") == "resolved":
                        resolved += 1
                    elif event.get("result") == "failed":
                        failed += 1
        signals["evidence"]["uc_with_evidence"] = with_evidence
        signals["healing"] = {
            "total_events": total_healing,
            "resolved": resolved,
            "failed": failed,
        }

    # Baseline for tests total/passed
    baseline_dir = quality_dir / "baselines"
    if baseline_dir.is_dir():
        for bf in baseline_dir.glob("*.json"):
            try:
                data = json.loads(bf.read_text(encoding="utf-8"))
            except (OSError, json.JSONDecodeError):
                continue
            metrics = data.get("metrics", {})
            signals["tests"]["total"] = int(metrics.get("tests_total", signals["tests"]["total"]))
            signals["tests"]["passed"] = int(metrics.get("tests_passed", signals["tests"]["passed"]))

    # Acceptance + AC + board from state registry (if state_path provided)
    if state_path:
        project_state = state_path / "projects" / project_name
        ac_file = project_state / "ac_status.json"
        if ac_file.exists():
            try:
                data = json.loads(ac_file.read_text(encoding="utf-8"))
                signals["ac_status"] = {
                    "total": int(data.get("total", 0)),
                    "completed": int(data.get("completed", 0)),
                }
            except (OSError, json.JSONDecodeError):
                pass
        board_file = project_state / "board_status.json"
        if board_file.exists():
            try:
                data = json.loads(board_file.read_text(encoding="utf-8"))
                signals["board"] = {
                    "us_total": int(data.get("us_total", 0)),
                    "us_blocked": int(data.get("us_blocked", 0)),
                }
            except (OSError, json.JSONDecodeError):
                pass
        acc_file = project_state / "acceptance_summary.json"
        if acc_file.exists():
            try:
                data = json.loads(acc_file.read_text(encoding="utf-8"))
                signals["acceptance"] = {
                    "accepted": int(data.get("accepted", 0)),
                    "rejected": int(data.get("rejected", 0)),
                }
            except (OSError, json.JSONDecodeError):
                pass

    return signals

File: .quality/scripts/audit/lib/signals.mjs
/**
 * lib/signals.mjs — Client-side SpecBox signals (UC-663).
 *
 * The Python server/audit/signals.py read these from the engine state
 * registry on the MCP host (the VPS in remote setups) — wrong filesystem.
 * This client version reads everything from the LOCAL repo:
 *
 *   - ac_status / board     ← doc/tracking/items.json (FreeForm source of truth)
 *   - evidence / healing    ← .quality/evidence/<feature>/{*, healing.jsonl}
 *   - tests                 ← .quality/baselines/*.json
 *   - acceptance            ← derived from board AC + review states
 *
 * Never throws — returns the neutral default block on any failure, exactly
 * like the Python (so analyzers degrade gracefully when a project has no
 * board or no .quality tree).
 */

import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { join } from 'node:path';

/** Neutral defaults — identical keys to signals.py's `signals` dict. */
function defaults() {
  return {
    ac_status: { total: 0, completed: 0 },
    evidence: { uc_total: 0, uc_with_evidence: 0 },
    healing: { total_events: 0, resolved: 0, failed: 0 },
    board: { us_total: 0, us_blocked: 0 },
    acceptance: { accepted: 0, rejected: 0 },
    tests: { total: 0, passed: 0 },
    prd_divergence_ratio: 0.0,
  };
}

/** Safe JSON parse of a file, or null. */
function readJson(absPath) {
  try {
    return JSON.parse(readFileSync(absPath, 'utf-8'));
  } catch {
    return null;
  }
}

/**
 * Gather SpecBox signals for a project from the local filesystem.
 *
 * @param {string} root - absolute project root (git toplevel)
 * @returns {object} signals dict matching signals.py shape
 */
export function fetchSpecboxSignals(root) {
  const signals = defaults();

  collectBoardSignals(root, signals);
  collectEvidenceAndHealing(root, signals);
  collectTests(root, signals);

  return signals;
}

/**
 * AC / US / acceptance counts from doc/tracking/items.json.
 *
 * FreeForm items carry `labels` (e.g. ["UC"], ["AC"], ["US"]) and a `state`
 * (done / review / todo / backlog / ...). We count AC items for ac_status,
 * US items for board, and treat AC in a rejected/blocked state as acceptance
 * rejections. This is a heuristic equivalent of the server's per-project
 * state files, derived from the single local source of truth.
 */
function collectBoardSignals(root, signals) {
  const itemsPath = join(root, 'doc', 'tracking', 'items.json');
  if (!existsSync(itemsPath)) return;
  const items = readJson(itemsPath);
  if (!Array.isArray(items)) return;

  const hasLabel = (it, label) =>
    Array.isArray(it.labels) && it.labels.includes(label);
  const tipoOf = (it) => (it.meta && it.meta.tipo) || null;
  const isAc = (it) => hasLabel(it, 'AC') || tipoOf(it) === 'AC';
  const isUs = (it) => hasLabel(it, 'US') || tipoOf(it) === 'US';

  let acTotal = 0;
  let acDone = 0;
  let acRejected = 0;
  let usTotal = 0;
  let usBlocked = 0;

  for (const it of items) {
    const state = it.state || '';
    if (isAc(it)) {
      acTotal += 1;
      if (state === 'done') {
        acDone += 1;
      } else if (state === 'rejected' || state === 'blocked') {
        acRejected += 1;
      }
    } else if (isUs(it)) {
      usTotal += 1;
      if (state === 'blocked') usBlocked += 1;
    }
  }

  signals.ac_status = { total: acTotal, completed: acDone };
  signals.board = { us_total: usTotal, us_blocked: usBlocked };
  signals.acceptance = { accepted: acDone, rejected: acRejected };
}

/**
 * Evidence + healing from .quality/evidence/<feature>/.
 * Mirrors signals.py: a feature dir with any file counts as "has evidence";
 * healing.jsonl lines with result resolved/failed are tallied.
 */
function collectEvidenceAndHealing(root, signals) {
  const evidenceDir = join(root, '.quality', 'evidence');
  if (!isDir(evidenceDir)) return;

  let features;
  try {
    features = readdirSync(evidenceDir, { withFileTypes: true })
      .filter((e) => e.isDirectory())
      .map((e) => join(evidenceDir, e.name));
  } catch {
    return;
  }

  let withEvidence = 0;
  let totalHealing = 0;
  let resolved = 0;
  let failed = 0;

  for (const dir of features) {
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }
    if (entries.some((e) => e.isFile())) withEvidence += 1;

    const healingFile = join(dir, 'healing.jsonl');
    if (existsSync(healingFile)) {
      const text = safeRead(healingFile);
      for (const line of text.split('\n')) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        let event;
        try {
          event = JSON.parse(trimmed);
        } catch {
          continue;
        }
        totalHealing += 1;
        if (event.result === 'resolved') resolved += 1;
        else if (event.result === 'failed') failed += 1;
      }
    }
  }

  signals.evidence = { uc_total: features.length, uc_with_evidence: withEvidence };
  signals.healing = { total_events: totalHealing, resolved, failed };
}

/** tests total/passed from .quality/baselines/*.json (mirrors signals.py). */
function collectTests(root, signals) {
  const baselineDir = join(root, '.quality', 'baselines');
  if (!isDir(baselineDir)) return;
  let files;
  try {
    files = readdirSync(baselineDir).filter((f) => f.endsWith('.json'));
  } catch {
    return;
  }
  for (const f of files) {
    const data = readJson(join(baselineDir, f));
    if (!data) continue;
    const metrics = data.metrics || {};
    if (metrics.tests_total !== undefined) {
      signals.tests.total = Number(metrics.tests_total) || signals.tests.total;
    }
    if (metrics.tests_passed !== undefined) {
      signals.tests.passed = Number(metrics.tests_passed) || signals.tests.passed;
    }
  }
}

function isDir(absPath) {
  try {
    return statSync(absPath).isDirectory();
  } catch {
    return false;
  }
}

function safeRead(absPath) {
  try {
    return readFileSync(absPath, 'utf-8');
  } catch {
    return '';
  }
}
